### ISSUE TITLE
v3.2.0 Release discussion

### ISSUE THREAD

**Issue by joshxg** [openmsupply/mobile#1675]
## RELEASE VERSION

v3.2.0 

### SUMMARY

Aim is to have all issues completed by January 7th, with testing completed and a final merge by January 14th.

This releases main goal is to integrate the new regimen data/indicator feature which has been introduced recently in desktop into mobile (Epic coming soon!) as well as to finally get the mobile dashboard into master!

There will also be various new bug fixes and refactors. Please use this issue as a forum for discussing issues that should be merged into master with `v3.2.0`

### Features

- [x] Mobile Dashboard: https://github.com/openmsupply/mobile/issues/1705
- [x] Program Indicators: https://github.com/openmsupply/mobile/issues/1685
- [x] Localization import/export: #484 
- [x] Messages: https://github.com/openmsupply/mobile/issues/1849

### Refactors

- [x] #1547
- [x] #1545
- [x] #1542
- [x] #1541
- [x] #1540
- [x] #1042 
- [x] #1404

### Bugs

- [x] #1435
- [x] #1777 

### DOCUMENTATION
- [x]  Dashboard
    - [Draft](https://docs.google.com/document/d/1UwusfNUIdbrRV7WIKwgEknVdMXskhu5ZfWeY4aKrhGU/edit?usp=sharing)
  - [Final] TODO.
- [x]  Indicators
    - [Draft](https://docs.google.com/document/d/1GxbT1cFJiLqzfuao1387oxrYXJHA4noub3nw2fGGHqs/edit?usp=sharing)
  - [Final] TODO.

**Comment by joshxg** [comment]
We already have PR's for the following:

- #976 
- #1667 
- #1608 
- #484 

So going to add these to the list.

----- 

## Refactoring: 

There is quite a lot of additional refactoring issues, so would be best to target a few. I'm going to suggest modals, which would be the following issues:

- #1547 
- #1545 
- #1543 
- #1542 
- #1541 
- #1540 

I have left #1539  and #1544 off this list as they are more pages rather than simple modals.

## Existing Bugs:

I think it would also be very beneficial to target these bugs to be fixed:

- #1435 
- #1401 
- #1477 
- #1474 
- #1434 

## Features

I think indicators and dashboard are plenty of features to add - I don't think any new features should be added unless specifically requested by a client

**Comment by Chris-Petty** [comment]
I agree with these lists more or less. ETA 3.2.0 though? Next week is my last week for the year, then January where we ramp up on dispensing 😕 @Stanzzer @craigdrown

**Comment by joshxg** [comment]
I put the milestone as around January 20th as a placeholder, but I think that would work. We need the dashboard and indicators for Ivory Coast regardless. The minor refactoring and bug fixes are more chores and house-keeping to keep the backlog at bay. We need a list of things to do and a time to have it done by - this does not include dispensing which would be 4.0.0

Potentially indicators will be non-backwrads compatiable so would need to bump the major when that is merged with master

**Comment by Chris-Petty** [comment]
I think that's a fair milestone, just being conscious of when dispensing is gonna happen

**Comment by diegoevangelisti** [comment]
If you guys agree, I would like to start working on those refactoring issues. (At least give it a try).

Issues #1608, #484 are related and waiting for review (some extra work associated with that PR can occur, depending on reviewer's feedback).

Same as always, feel free to assign me to any issue you believe we need to prioritize.

**Comment by joshxg** [comment]
> If you guys agree, I would like to start working on those refactoring issues. (At least give it a try).

That is what I'm thinking! It's a great learning opportunity and quite isolated from other work that might be occurring

**Comment by Chris-Petty** [comment]
@diegoevangelisti Do iiittttt

**Comment by joshxg** [comment]
Cool, looks like the list is a good ahead?

@wlthomson - I guess you're on indicators, maybe we can get some issues started soon to begin discussion

We can also look at getting the dashboard PR updated and merged - we might need to consider new options for the backend and potentially adjust accordingly, but the basic page looks like it is in alright state to start iterations?

**Comment by Stanzzer** [comment]
@joshxg @Chris-Petty @diegoevangelisti @wlthomson and others, thanks for getting this underway and allocating/picking up some tasks. It would be good if we can agree a timeline and task allocation to ensure we are doing our best to meet the delivery date in late Jan.

**Comment by wlthomson** [comment]
> Cool, looks like the list is a good ahead?
> 
> @wlthomson - I guess you're on indicators, maybe we can get some issues started soon to begin discussion
> 
> We can also look at getting the dashboard PR updated and merged - we might need to consider new options for the backend and potentially adjust accordingly, but the basic page looks like it is in alright state to start iterations?

@joshxg @Chris-Petty Do you have a time that works well for you that we could do a group skype?

**Comment by joshxg** [comment]
@wlthomson indicators being one of the big features might need to dictate the timeline here a little. Could we get a ball park on how long you're thinking? Maybe we can get a discussion going after we have some issues made and can also tie that in?

**Comment by wlthomson** [comment]
> @wlthomson indicators being one of the big features might need to dictate the timeline here a little. Could we get a ball park on how long you're thinking? Maybe we can get a discussion going after we have some issues made and can also tie that in?

Sounds good. I'll make some issues and wireframes up today/this weekend and link here. I think there will be sufficient stuff here to subdivide into more than one release (maybe we could make use of minor versions for the refactoring stuff etc.). If we could divide all the IC mobile tasks due late Jan into a subset of milestones, that might be helpful to track progress.

**Comment by joshxg** [comment]
This is just for the `3.2` release. The above list which is really just a few bug fixes and a couple small refactors, as well as indicators and possibly the dashboard pr. All in all this is a small amount of work.

I think we should just keep it simple and just have this simple list and an indicators epic and we can improve on this with later releases.

**Comment by wlthomson** [comment]
> This is just for the `3.2` release. The above list which is really just a few bug fixes and a couple small refactors, as well as indicators and possibly the dashboard pr. All in all this is a small amount of work.
> 
> I think we should just keep it simple and just have this simple list and an indicators epic and we can improve on this with later releases.
> 
> If we could get that ball park amount of time from you that would be helpful.

OK, but to me it seems like mixing everything together into one release doesn't make things simpler and muddles visibility on feature progress, even for small releases. The gitflow approach of separating bugs and refactors into a minor releases and features (i.e. indicators and dashboard) is simpler IMO. But fine to leave as is.

Re. time estimate, sorry if I wasn't clear. I want to revisit desktop indicators and have a quick think about how the design can be done on mobile and make some issues so I can scope out the things that need to be done properly before I can give an estimate that means anything. If I gave a ballpark figure now I'd just be guessing and would probably have to revise it on Monday 🤷‍♂ .

**Comment by wlthomson** [comment]
Fair enough.

Re. time estimate, sorry if I wasn't clear. I want to revisit desktop indicators and have a quick think about how the design can be done on mobile and make some issues. I think it's a good idea for me to scope out the things that need to be done a bit before I give an estimate. I'm reluctant to give a ballpark figure just now as there are a few variables and I'd be mostly guessing and would probably have to revise it.

If I had to guess, I'd estimate somewhere in the 1-2 weeks. If we can lockdown the desktop implementation (I'm still a bit worried that it hasn't had much feedback), then that would definitely help speed things up.

**Comment by joshxg** [comment]
Cool - I say we aim for the 7th of Jan then, to get this list done with a week of testing till the 14th of Jan?

**Comment by wlthomson** [comment]
Sounds good. I think re. indicators the first steps will be locking down some front-end design decisions. That way we can get moving while waiting for feedback on the desktop implementation and decisions re. sync etc. 

Will get started on the epic and get the ball rolling.

**Comment by joshxg** [comment]
We've got our two main features with epics now: #1685 and #1705

Also a new bug being added to the list of things to fix with this release: https://github.com/openmsupply/mobile/issues/1680

**Comment by joshxg** [comment]
Quick update:

Dashboard is progressing nicely and can be tracked here: https://github.com/openmsupply/mobile/projects/8 

The related desktop issue can be seen here: https://github.com/sussol/msupply/issues/2179

----- 

Bugs and refactors are moving along.

- I have had a look at #1435 and think I might have a solution, will need to test on a device though.
- #1474 and #1477 however I cannot recreate or figure out what is wrong if anyone is able to have a look!

**Comment by joshxg** [comment]
- #1435 turned out to not be a problem! The instruction set on the device tested on was different than expected, so when using the correct APK it worked fine!

- Still unsure on #1474 and #1474 though!

- Indicators are moving along [here](https://github.com/openmsupply/mobile/issues/1685) -where wireframes have been mocked up for the implementation

**Comment by Rebecca367** [comment]
So behind on this issue. @joshxg thanks for getting me involved in developing status even t it's behind. aim to finish testing cases as much as possible to be productive before going on holidays. Aim for another release right after holiday season is quite ambitious, isn't it?!

**Comment by joshxg** [comment]
- Have had a little bit of a hiccup with the merge of #484 but has been resolved with #1772 (hopefully). Will need to have a bit more testing done on this unfortunately! 
- Refactor issues are complete, with the bugs still needing to have a look at if anyone has the time.
- Dashboard frontend will slow down a bit as diego is on break, but will aim to keep that moving forward as it is almost done with the backend in desktop.
- Indicators also slowing as will is travelling, will have to evaluate if this might get bumped to the next release in the next few days!

-----

We've also had returns come up as a feature to be implemented in mobile: https://github.com/openmsupply/mobile/issues/1770 -- also a feature most likely to be added to the next release, although depending on the implementation, might be simple enough to push into this one!

**Comment by joshxg** [comment]
Quick update: 

- Anything that hasn't been done (bugs) are going to go on the backburner
- Indicators is looking solid, probably looking to get done in the next couple of days and tested. Mainly needing to do migration code handling.
- Dashboard is basically done with just some localization to complete. Just waiting on desktop backend to be completed before merging this in.

For both dashboard and indicators, it looks like we are going to need to have some system for re-syncing tables: https://github.com/openmsupply/mobile/issues/1849 - so going to add this into this release. This means the dates are going to get pushed a little, but if we start getting testing done now, we shouldn't be too far off at all!

**Comment by joshxg** [comment]
Update:

Have just released the first release candidate for testing purposes. This does NOT contain indicators or dashboard yet.. hopefully soon! :) 🎉

**Comment by joshxg** [comment]
Update: 

Bit of testing was able to get done so far on the RC version.. thanks @diegoevangelisti !

- Should be looking at merging the dashboard feature branch today after a little bit of testing before merging.
- Indicators has a couple of issues remaining before it can be merged.. looking good to be in early next week.
- Have had to add a new issue to this release to help with indicator migration: https://github.com/openmsupply/mobile/issues/1849

**Comment by Gwalior-84** [comment]
@joshxg I had understood that patient dispensing would be in this version.  It doesn't appear in the Features list above.  Could you please confirm what the plan is with Patient dispensing?

**Comment by joshxg** [comment]
@Gwalior-84 Dispensing is a little larger, so the plan is to split these all up and have dispensing in the next version. Hopefully in about 2 weeks.

**Comment by joshxg** [comment]
Update:

Unfortunately we didn't get dashboard merged in last week. We are just tidying up some UI changes - specifically labels before we merge this into develop.

Regarding indicators - there is just an issue regarding outgoing sync that we need to address, then we can merge that into develop for some testing!

Looking like we might be a little pushed back, but not too much!

**Comment by joshxg** [comment]
Update:

- Have merged Dashboard and planning to give that a good test today.
- Last PR for indicators is in, just need to do an update from develop and we will merge that and give that a good test today also!

Looking promising to have a release out very soon pending any issues!

**Comment by joshxg** [comment]
Update:

- Have tested dashboard and that has gone really well! Final thing is a little bit of draft documentation from @diegoevangelisti :)
- Indicators has been merged and test cases/documentation is currently being written by @wlthomson 

All going well, we should be releasing tomorrow!

**Comment by wlthomson** [comment]
Update:

- Indicators test cases (see #1935) and documentation is completed 😄 .

**Comment by joshxg** [comment]
@wlthomson @diegoevangelisti 

- Feel free to edit the top comment with your draft docs once they're done!

**Comment by diegoevangelisti** [comment]
@joshxg 
The Draft doc for Dashboard is ready and it was added on top :-)

**Comment by wlthomson** [comment]
Update:

Due to some indicators sync problems delaying testing and some additional indicators fixes which I didn't manage to get done today, it looks like we will be pushing the release over the weekend to Monday 😢 .

That should give us the remaining time to get the remaining fixes merged and tested, and hopefully provide a bit of extra time for some additional feature additions or tidy ups on dispensing or dashboard.

**Comment by joshxg** [comment]
Update:

Couple more bugs to be fixed with indicators before we're able to release. Hopefully tomorrow.

**Comment by wlthomson** [comment]
Update:

Indicators functionality completed and all PRed. Working with @diegoevangelisti to complete the indicators testing.

Subject to testing, we should be good to release this afternoon (or first thing tomorrow if @joshxg laptop needs more time to dry out 😢 ).

**Comment by joshxg** [comment]
3.2.0 is released 🎉 🎉 

We're just waiting on the final documentation from @loulomas (?) so will keep this open until then. 

Feel free to check out the new issues for further discussions!

- For the 3.3.0 release: #1978
- For patch releases for 3.2.X releases: #1979

**Comment by wlthomson** [comment]
Update:

- v3.2.0 release candidate merged (#1977).
- v3.2.0 released (`releases/tag/v3.2.0`).
- Closing and migrating remaining documentation tasks to patch release discussion.

**Comment by Gwalior-84** [comment]
Thanks for this @joshxg .  Please remind me how to get to the APK...  I've spent five minutes looking for it, but in vain.

**Comment by joshxg** [comment]
@Gwalior-84 the apks are here: https://github.com/openmsupply/mobile/releases -- the universal APK is for the general use where as the more specific APK's are smaller but require you to know exactly what the CPU architecture and instruction set are

**Comment by LouLomas** [comment]
@joshxg @diegoevangelisti @wlthomson - still just two drafts attached, are you working on anymore?

**Comment by diegoevangelisti** [comment]
> @joshxg @diegoevangelisti @wlthomson - still just two drafts attached, are you working on anymore?

My draft is done (Dashboard). I guess just Indicators and Dashboard docs are needed for this release, so if I'm not wrong that would be all Lou.
